Refactor linear solver to handle symmetric and complex inputs

gcmartins opened the pull request:
This is part of decoupling harmonic and modal solver efforts. Other PRs will be open to handle other parts of it.
This should work the same way as before and can be merged independently of the other parts of the work.

gcmartins committed:
Refactor linear solver to handle symmetric and complex inputs

JacsonGilVargas reviewed (changes requested):
I've performed a few tests with acoustic analyses and found a problem while accessing the complex acoustic modal analysis data. The same problem doesn't appear in the non-complex acoustic modal analysis.


<img width="542" height="319" alt="image" src="https://github.com/user-attachments/assets/53858148-737e-475d-9a05-15d29afe50e5" />


The structural harmonic analyses is showing inconsistent results, even for undamped problems.


<img width="1235" height="798" alt="image" src="https://github.com/user-attachments/assets/aed64b00-147b-40d8-bec4-bf1468f2676c" />


What we should expect as a consistent result for the tested model at 325Hz, for instance.


<img width="1231" height="789" alt="image" src="https://github.com/user-attachments/assets/5eb1677d-3e61-4960-9558-8c344c16d218" />


Finally, structural modal analysis also yields physically inconsistent results. Fortunately, the dev branch is working well and can be used as a reference for addressing issues related to global matrices.

gcmartins commented:
> I've performed a few tests with acoustic analyses and found a problem while accessing the complex acoustic modal analysis data. The same problem doesn't appear in the non-complex acoustic modal analysis.
> 
> <img alt="image" width="542" height="319" src="https://private-user-images.githubusercontent.com/57151261/485841341-53858148-737e-475d-9a05-15d29afe50e5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyODUwMjksIm5iZiI6MTc1NzI4NDcyOSwicGF0aCI6Ii81NzE1MTI2MS80ODU4NDEzNDEtNTM4NTgxNDgtNzM3ZS00NzVkLTlhMDUtMTVkMjlhZmU1MGU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA3VDIyMzg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwOTk2Y2ExNGZjZjIyY2I0NTJjNzIwNmFhMDhkYjJhMzViMDgxNDA5YzM5MDQ0ZWVlOTRiMzY1ZmUwM2U1Y2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P2Dsyd7MHUj9ZtmfD63W4OZakIpZJBolQSY9LpwIlVE">
> The structural harmonic analyses is showing inconsistent results, even for undamped problems.
> 
> <img alt="image" width="1235" height="798" src="https://private-user-images.githubusercontent.com/57151261/485847464-aed64b00-147b-40d8-bec4-bf1468f2676c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyODUwMjksIm5iZiI6MTc1NzI4NDcyOSwicGF0aCI6Ii81NzE1MTI2MS80ODU4NDc0NjQtYWVkNjRiMDAtMTQ3Yi00MGQ4LWJlYzQtYmYxNDY4ZjI2NzZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA3VDIyMzg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3Y2VjZGU1YjE2MDBiYThhODQyODYzM2E4MDMyYjA3ZTBmMDBmZDcwMDcyYjE5NjRjMGNhY2M0MWQ5YzZlZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VsIlo3Cc3czd2i5XYgfq85k1rj7vm8AelSfziENu304">
> What we should expect as a consistent result for the tested model at 325Hz, for instance.
> 
> <img alt="image" width="1231" height="789" src="https://private-user-images.githubusercontent.com/57151261/485848884-5eb1677d-3e61-4960-9558-8c344c16d218.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyODUwMjksIm5iZiI6MTc1NzI4NDcyOSwicGF0aCI6Ii81NzE1MTI2MS80ODU4NDg4ODQtNWViMTY3N2QtM2U2MS00OTYwLTk1NTgtOGMzNDRjMTZkMjE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA3VDIyMzg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4NGYwMzA5YWVlMDlhY2RlZjVhMmQ4YjEyNjVhODVjNmZhZmZmNjM4ZGU3MGJkNWEzNjQ2ZWI5NmU4NmE4ZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L4QsWSZYqezjxLIxwo25zZz3G-2x1cof9J44Kwr-2C8">
> Finally, structural modal analysis also yields physically inconsistent results. Fortunately, the dev branch is working well and can be used as a reference for addressing issues related to global matrices.

I've found two issues while investigating this problem:

- Since we're now detecting symmetric matrices, the structural matrix is ​​showing some strange non-symmetric entries, I don't know why. Even in a free-form condition, it still has non-symmetric entries. I think we should investigate this.
- This non-symmetric matrix shouldn't be the main cause of the strange result you found. PARDISO supports non-symmetric matrices, but I don't think it's working well in this case. If you switch to the MUMPS solver, the results will be practically the same as before.

The results were consistent before because we were assuming the matrices were symmetric instead of detecting them and passing the symmetric matrix flag to PARDISO.

gcmartins committed:
Putting back pre-convertion to real matrices when it is applicable

gcmartins commented:
> I noticed that only real non-symmetric solver (PARDISO) is giving weird results, so I change the logic to enforce complex non-symmetric solver for any non-symmetric matrix.


I think the acoustic modal analysis issue is due to this same reason. The PARDISO real non-symmetric solver is giving weird results.

gcmartins committed:
Move postprocessing and assembly related things out of AcousticHarmonicSolver

gcmartins committed:
Move postprocessing and assembly related things from AcousticModalSolver

gcmartins committed:
Move postprocessing and assembly related things from StructuralHarmonicSolver

gcmartins committed:
Unify modal solver

gcmartins committed:
Unify harmonic solver

gcmartins committed:
Refactor to include saving by frequency approach on the structural harmonic solver

gcmartins committed:
Validation test adjustments

gcmartins committed:
Implement approach to set matrix symmetry assumption

gcmartins committed:
Fix renamed method

JacsonGilVargas committed:
Rename the solve methods to solve_direct in acoustic validation scripts

andrefpf reviewed (approved):
As long as I tested everything seems to be working fine.

gcmartins committed:
Fix validate test

JacsonGilVargas reviewed (approved):
I've conducted several tests in both modal and harmonic analysis in the acoustic and structural physical domains, and everything worked very well. The proposed refactorization reorganized the code, with great emphasis on the solvers. Excellent work @gcmartins, congratulations.

gcmartins committed:
Solving conflicts: replacing dofs -> dof

gcmartins committed:
Small fix for backward compatibility
